File: server/src/services/rag/reranker_service.py
"""
Reranker Microservice - BGE-reranker-v2-m3
Phase 1 - RAG V2 Implementation

Usage: uvicorn reranker_service:app --host 0.0.0.0 --port 8001
"""

import os
import time
from typing import List, Optional
from contextlib import asynccontextmanager

from fastapi import FastAPI, HTTPException
from pydantic import BaseModel
from sentence_transformers import CrossEncoder

# Configuration
MODEL_NAME = os.getenv("RERANKER_MODEL", "BAAI/bge-reranker-v2-m3")
DEFAULT_TOP_K = int(os.getenv("RERANKER_TOP_K", "5"))

# Global model instance
model: Optional[CrossEncoder] = None


class RerankRequest(BaseModel):
    query: str
    documents: List[str]
    top_k: Optional[int] = None


class RerankResult(BaseModel):
    index: int
    score: float
    document: str


class RerankResponse(BaseModel):
    results: List[RerankResult]
    processing_time_ms: float
    model: str


class HealthResponse(BaseModel):
    status: str
    model: str
    ready: bool


@asynccontextmanager
async def lifespan(app: FastAPI):
    """Load model on startup."""
    global model
    print(f"Loading reranker model: {MODEL_NAME}")
    start = time.time()
    model = CrossEncoder(MODEL_NAME, max_length=512)
    print(f"Model loaded in {time.time() - start:.2f}s")
    yield
    print("Shutting down reranker service")


app = FastAPI(
    title="Reranker Service",
    description="BGE-reranker-v2-m3 for RAG reranking",
    version="1.0.0",
    lifespan=lifespan
)


@app.get("/health", response_model=HealthResponse)
async def health():
    """Health check endpoint."""
    return HealthResponse(
        status="ok" if model else "loading",
        model=MODEL_NAME,
        ready=model is not None
    )


@app.post("/rerank", response_model=RerankResponse)
async def rerank(request: RerankRequest):
    """
    Rerank documents for a given query.

    Returns documents sorted by relevance score (highest first).
    """
    if not model:
        raise HTTPException(status_code=503, detail="Model not loaded")

    if not request.documents:
        return RerankResponse(
            results=[],
            processing_time_ms=0,
            model=MODEL_NAME
        )

    start = time.time()

    # Create query-document pairs
    pairs = [[request.query, doc] for doc in request.documents]

    # Get scores from cross-encoder
    scores = model.predict(pairs)

    # Create results with original indices
    results = [
        RerankResult(
            index=i,
            score=float(scores[i]),
            document=doc
        )
        for i, doc in enumerate(request.documents)
    ]

    # Sort by score descending
    results.sort(key=lambda x: x.score, reverse=True)

    # Apply top_k limit
    top_k = request.top_k or DEFAULT_TOP_K
    results = results[:top_k]

    processing_time = (time.time() - start) * 1000

    return RerankResponse(
        results=results,
        processing_time_ms=round(processing_time, 2),
        model=MODEL_NAME
    )


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8001)

File: server/src/services/rag/RerankerService.ts
/**
 * RerankerService - Node.js client for Python Reranker Microservice
 * Phase 1 - RAG V2 Implementation
 *
 * Model: BGE-reranker-v2-m3 (Multilingual)
 */

export interface RerankerConfig {
  serviceUrl: string;
  topK: number;
  enabled: boolean;
  timeout: number;
}

export interface ScoredChunk {
  id: string;
  content: string;
  score: number;
  documentId: string;
  chunkIndex: number;
  metadata?: Record<string, unknown>;
}

export interface RerankedChunk extends ScoredChunk {
  rerankerScore: number;
  originalRank: number;
  newRank: number;
}

export interface RerankerResult {
  chunks: RerankedChunk[];
  processingTimeMs: number;
  modelUsed: string;
}

interface RerankApiResponse {
  results: Array<{
    index: number;
    score: number;
    document: string;
  }>;
  processing_time_ms: number;
  model: string;
}

interface HealthResponse {
  status: string;
  model: string;
  ready: boolean;
}

export class RerankerService {
  private config: RerankerConfig;
  private available: boolean = false;

  constructor(config?: Partial<RerankerConfig>) {
    this.config = {
      serviceUrl: config?.serviceUrl || process.env.RERANKER_SERVICE_URL || 'http://localhost:8001',
      topK: config?.topK || parseInt(process.env.RERANKER_TOP_K || '5'),
      enabled: config?.enabled ?? (process.env.RERANKER_ENABLED === 'true'),
      timeout: config?.timeout || 30000, // 30s for up to 20 documents
    };
  }

  /**
   * Check if reranker service is available
   */
  async checkHealth(): Promise<boolean> {
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 3000);

      const response = await fetch(`${this.config.serviceUrl}/health`, {
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (response.ok) {
        const health: HealthResponse = await response.json();
        this.available = health.ready;
        return health.ready;
      }
      return false;
    } catch {
      this.available = false;
      return false;
    }
  }

  /**
   * Initialize the service - check availability
   */
  async initialize(): Promise<void> {
    if (!this.config.enabled) {
      console.log('Reranker: Disabled by configuration');
      return;
    }

    const healthy = await this.checkHealth();
    if (healthy) {
      console.log(`Reranker: Connected to ${this.config.serviceUrl}`);
    } else {
      console.warn(`Reranker: Service not available at ${this.config.serviceUrl}`);
    }
  }

  /**
   * Rerank chunks based on query relevance
   */
  async rerank(query: string, chunks: ScoredChunk[], topK?: number): Promise<RerankerResult> {
    // If disabled or unavailable, return original order
    if (!this.config.enabled || !this.available) {
      return {
        chunks: chunks.slice(0, topK || this.config.topK).map((c, i) => ({
          ...c,
          rerankerScore: c.score,
          originalRank: i,
          newRank: i,
        })),
        processingTimeMs: 0,
        modelUsed: 'none',
      };
    }

    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.config.timeout);

      const response = await fetch(`${this.config.serviceUrl}/rerank`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({
          query,
          documents: chunks.map(c => c.content),
          top_k: topK || this.config.topK,
        }),
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        throw new Error(`Reranker API error: ${response.status}`);
      }

      const data: RerankApiResponse = await response.json();
      console.log(`🔧 DEBUG RerankerService: API returned ${data.results.length} results`);

      // Map results back to chunks with reranker scores
      const rerankedChunks: RerankedChunk[] = data.results.map((result, newRank) => {
        const originalChunk = chunks[result.index];
        return {
          ...originalChunk,
          rerankerScore: result.score,
          originalRank: result.index,
          newRank,
        };
      });

      return {
        chunks: rerankedChunks,
        processingTimeMs: data.processing_time_ms,
        modelUsed: data.model,
      };
    } catch (error) {
      console.error('Reranker error:', error);
      // Fallback to original order on error
      return {
        chunks: chunks.slice(0, topK || this.config.topK).map((c, i) => ({
          ...c,
          rerankerScore: c.score,
          originalRank: i,
          newRank: i,
        })),
        processingTimeMs: 0,
        modelUsed: 'fallback',
      };
    }
  }

  /**
   * Check if reranking is enabled and available
   */
  isAvailable(): boolean {
    return this.config.enabled && this.available;
  }

  /**
   * Get current configuration
   */
  getConfig(): RerankerConfig {
    return { ...this.config };
  }
}

// Singleton instance
export const rerankerService = new RerankerService();
